cbv3_django_prototype/cbv3_django_prototype/resources/models.py

```diff
@@ -1,3 +1,47 @@
+import uuid
+import django.contrib.postgres.fields as pg
 from django.db import models
+from django.utils import timezone
 
-# Create your models here.
+class Resource(models.Model):
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid1, editable=False)
+
+    title = models.CharField(max_length=200)
+
+    #potentially a markdown field.  Will need a markdown converter and renderer
+    description = models.TextField(blank=True, max_length=500)
+
+    #specific URL of resource
+    url = models.URLField(max_length=300)
+
+    #the URL of the referring/source site.  e.g. URL of tweet, if it was tweeted.
+    referrer = models.URLField(blank=True, max_length=300)
+
+    #names of persons who suggest this resource to the user
+    credit = models.CharField(max_length=100)
+
+    # publication date of resource
+    date_published = models.DateTimeField(default=timezone.now)
+
+    # creation date of resource entry
+    created = models.DateTimeField(auto_now_add=True)
+
+    # modification date of resource entry
+    modified = models.DateTimeField(default=timezone.now)
+
+    # [ video, podcast, talk, tutorial, course, book, blog ]
+    type = models.CharField(max_length=100)
```

Comment:
These options can be constrained with a `CHOICES` argument.

Reply:
++ Those options in the comment look pretty comprehensive to me, fwiw.

Reply:
Since the front end will be using React, I thought the constraint might be better there.  Also gives some flexibility if (and when!) the fields change....because they will...no matter how comprehensive they look now.  'Game' is one addition that comes to mind.  ;)

Reply:
Yeah, I think the backend should have the final say in validation. Adding choices constraint might also allow DRF to auto-document the allowed values.

Reply:
I guess I am not clear on how a CHOICES constraint works when we will not be using any front-end templates or views and will have all front end code in React.  But I will add on next update.

Reply:
I believe the `choices` would be used to validate the input data:

> If choices are given, they’re enforced by [model validation](https://docs.djangoproject.com/en/2.2/ref/models/instances/#validating-objects)
>
> https://docs.djangoproject.com/en/2.2/ref/models/fields/#choices

Also, we want to decouple the frontend from the server (as other clients may emerge with something like Vue, Flutter, etc), so the server should be able to stand alone -- including correctly validating data.

Reply:
It appears to be a little more complicated than that...at least from my reading today.  But I am taking a look at it.  https://stackoverflow.com/questions/28945327/django-rest-framework-with-choicefield

Reply:
Here are the configured **RESOURCE_TYPES:**
        [('VID', 'Video'),
        ('POD', 'Podcast'),
        ('PODEP', 'Podcast Episode'),
        ('TALK', 'Talk'),
        ('TUTOR', 'Tutorial'),
        ('COURSE', 'Course'),
        ('BOOK', 'Book'),
        ('BLOG', 'Blog'),
        ('GAME', 'Game'),
        ('EVENT', 'Event'),
        ('TOOL', 'Tool'),
        ('LIB', 'Library'),
        ('WEB', 'Website')]

The validation will be on the first value, the second will be what is displayed in a UI or given back.  The Django Amin shows a drop-down for these values, but the DRF admin omits the field entirely.